pkg/coscheduling/coscheduling.go

```diff
@@ -95,295 +80,316 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+func (cs *Coscheduling) VerifyLocalCache(toSchedulePGName string) bool {
+	pgStatus := cs.pgMgr.PodGroupStatusCache().Get(toSchedulePGName)
+	podInCacheUpdated := false
+	// verify pods cache
+	cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		klog.V(5).Info("Start IterateOverWaitingPods")
+		pod := waitingPod.GetPod()
+		podFullName := core.GetNamedspacedName(pod)
+		pgName := core.GetNamedspacedName(pgStatus.PodGroup)
+		klog.V(5).Infof("Start walking through pod %v uid: %v", podFullName, pod.UID)
+		if pgName != toSchedulePGName {
+			return
+		}
+		// Make sure pods in pgStatus cache are still exist. This may happen when pod is deleted and re-created.
+		klog.V(5).Infof("Wanted cache pod %v", podFullName)
+		_, found := pgStatus.AssumedPods.Get(podFullName)
+		if !found {
+			podInCacheUpdated = true
+			return
 		}
+		klog.V(6).Infof("Current pod group state %+v", pgStatus.AssumedPods.Items())
 	})
+	return podInCacheUpdated
 }
 
-// GetPodGroupLabels checks if the pod belongs to a PodGroup. If so, it will return the
-// podGroupName, minAvailable of the PodGroup. If not, it will return "" and 0.
-func GetPodGroupLabels(pod *v1.Pod) (string, int, error) {
-	podGroupName, exist := pod.Labels[PodGroupName]
-	if !exist || len(podGroupName) == 0 {
-		return "", 0, nil
-	}
-	minAvailable, exist := pod.Labels[PodGroupMinAvailable]
-	if !exist || len(minAvailable) == 0 {
-		return "", 0, nil
+// StartBatchSchedule receives the group name of pod to start scheduling
+func (cs *Coscheduling) StartBatchSchedule(fullName string) {
+	pgStatus := cs.pgMgr.PodGroupStatusCache().Get(fullName)
+	if pgStatus.PodGroup.Status.Phase != v1alpha1.PodGroupPreScheduling && pgStatus.PodGroup.Status.Phase != v1alpha1.
+		PodGroupScheduling {
+		return
 	}
-	minNum, err := strconv.Atoi(minAvailable)
+
+	// record time to avoid abnormal exit when bind
+	pg := pgStatus.PodGroup
+	pgCopy := pgStatus.PodGroup.DeepCopy()
+	pgCopy.Status.ScheduleStartTime = metav1.Now()
+
+	patch, err := util.CreateMergePatch(pg, pgCopy)
 	if err != nil {
-		klog.Errorf("PodGroup %v/%v : PodGroupMinAvailable %v is invalid", pod.Namespace, pod.Name, minAvailable)
-		return "", 0, err
+		return
 	}
-	if minNum < 1 {
-		klog.Errorf("PodGroup %v/%v : PodGroupMinAvailable %v is less than 1", pod.Namespace, pod.Name, minAvailable)
-		return "", 0, err
+
+	if err := cs.pgMgr.PatchPodGroup(pg.Name, pg.Namespace, patch); err != nil {
+		return
 	}
-	return podGroupName, minNum, nil
-}
 
-func (cs *Coscheduling) calculateTotalPods(podGroupName, namespace string) int {
-	// TODO get the total pods from the scheduler cache and queue instead of the hack manner.
-	selector := labels.Set{PodGroupName: podGroupName}.AsSelector()
-	pods, err := cs.podLister.Pods(namespace).List(selector)
-	if err != nil {
-		klog.Error(err)
-		return 0
+	// Start coscheduling
+	klog.V(4).Infof("Start scheduling %v", fullName)
+	localWaitingPods := cs.pgMgr.GetPodNodePairs(fullName)
```

Comment:
Why we need get the waiting pod from the nodepairs other than from the waiting pods map?